sdk/identity/azure-identity/src/chained_token_credential.cpp

```diff
@@ -4,69 +4,138 @@
 #include "azure/identity/chained_token_credential.hpp"
 
 #include "azure/core/internal/diagnostics/log.hpp"
+#include <azure/core/azure_assert.hpp>
 
 #include <utility>
 
 using namespace Azure::Identity;
 using namespace Azure::Core::Credentials;
 using Azure::Core::Context;
+using Azure::Core::Diagnostics::Logger;
+using Azure::Core::Diagnostics::_internal::Log;
+
+namespace {
+std::string const IdentityPrefix = "Identity: ";
+}
 
 ChainedTokenCredential::ChainedTokenCredential(ChainedTokenCredential::Sources sources)
-    : m_sources(std::move(sources))
+    : ChainedTokenCredential(sources, {}, {})
 {
 }
 
-ChainedTokenCredential::~ChainedTokenCredential() = default;
-
-AccessToken ChainedTokenCredential::GetToken(
-    TokenRequestContext const& tokenRequestContext,
-    Context const& context) const
+ChainedTokenCredential::ChainedTokenCredential(
+    ChainedTokenCredential::Sources sources,
+    std::string const& enclosingCredential,
+    std::vector<std::string> sourceDescriptions)
+    : m_sources(std::move(sources)), m_sourceDescriptions(std::move(sourceDescriptions))
 {
-  using Azure::Core::Diagnostics::Logger;
-  using Azure::Core::Diagnostics::_internal::Log;
+  // LCOV_EXCL_START
+  AZURE_ASSERT(m_sourceDescriptions.empty() || m_sourceDescriptions.size() == m_sources.size());
+  // LCOV_EXCL_STOP
 
-  auto n = 0;
-  for (const auto& source : m_sources)
+  auto const logLevel = m_sources.empty() ? Logger::Level::Warning : Logger::Level::Informational;
+  if (Log::ShouldWrite(logLevel))
   {
-    try
+    std::string credentialsList;
+    if (!m_sourceDescriptions.empty())
     {
-      ++n;
-      auto token = source->GetToken(tokenRequestContext, context);
-
+      auto const sourceDescriptionsSize = m_sourceDescriptions.size();
+      for (size_t i = 0; i < (sourceDescriptionsSize - 1); ++i)
       {
-        const auto logLevel = Logger::Level::Informational;
-        if (Log::ShouldWrite(logLevel))
-        {
-          Log::Write(
-              logLevel,
-              std::string("ChainedTokenCredential authentication attempt with credential #")
-                  + std::to_string(n) + " did succeed.");
-        }
+        credentialsList += m_sourceDescriptions[i] + ", ";
       }
 
-      return token;
+      credentialsList += m_sourceDescriptions.back();
     }
-    catch (const AuthenticationException& e)
+
+    Log::Write(
+        logLevel,
+        IdentityPrefix
+            + (enclosingCredential.empty()
```

Comment:
Enclosing credential is odd.  Why not log
  - DefaultCredential attempting to get Toke
  - ChainedToken attempting to create credential from source: <source friendly name>
  - ChainedToken attempting to create credential from source: <source friendly name>
  - ChainedToken successfully created credential from source: <source friendly name>
 
The fact DefaultAzureCredential wraps chainedtoken is implementation detail that chainedtoken shouldn't care about

Reply:
I thought about this, but decided to rather expose the fact that `ChainedTokenCredential` is used. I also disclose it in documentation. I am rather 55/45 to disclose. My thought is that by disclosing, we encourage/hint how users can create the very same credential as the next step or should they want to cusomize.

Reply:
#4419